src/main/java/org/testng/internal/TestInvoker.java

```diff
@@ -861,5 +721,80 @@ private static int computeTestStatusComparingTestResultAndStatusHolder(
     return testResult.getStatus();
   }
 
+  private class MethodInvocationAgent {
+
+    private ITestContext context;
+    private List<ITestResult> result = Lists.newArrayList();
+    private FailureContext failure = new FailureContext();
+    private AtomicInteger invocationCount;
+    private ITestInvoker invoker;
+    private final TestMethodArguments arguments;
```

Comment:
All should be final or only declared in the method

Reply:
Fixed. Except for `invocationCount` (which needs to get initialized everytime `invoke()` gets invoked) now everything is now final.

Reply:
Yes, but why is `invocationCount` declared in the class instead of the method?

Suggestion: initialize all attributes at the same place when possible (here the constructor). I find it more readable. At least, group construction styles together.

Reply:
Fixed.